File: 0_theory_of_computation/3_design_and_analysis_of_algorithms/common_computational_problems/0_declarative_computational_model/arrays/rotate_left.scala
import scala.annotation.tailrec
/*
    for rotations
        foreach index
            newIndex = index - 1
            if newIndex >= 0
                newArray[newIndex] = array[index]
            else
                newArray[lastIndex] = array[index]
 */
def rotateLeft(a: Array[Int], d: Int): Array[Int] = {

  // Example of calculating a future state once
  @tailrec
  def new_index(currentIndex: Int, rotationsLeft: Int): Int =
    val decrement = 1
    val newIndex = currentIndex - decrement
    if rotationsLeft <= 0 then currentIndex
    else 
      if newIndex < 0 then
        new_index(a.length - decrement, rotationsLeft - decrement)
      else new_index(newIndex, rotationsLeft - decrement)

  // def new_index(currentIndex: Int, rotations: Int) = 
  //   val newDistance = currentIndex - rotations
  //   if (newDistance < 0) newDistance.abs - 1 else newDistance

  val newArray = new Array[Int](a.length)

  a.zipWithIndex.foreach { case (value: Int, index: Int) =>
    newArray(new_index(index, d)) = value
  }

  newArray
}


@main def main(): Unit = 
  rotateLeft(Array(1, 2, 3, 4, 5), 3).foreach(print)
  assert(rotateLeft(Array(1, 2, 3, 4, 5), 3).sameElements(Array(4,5,1,2,3)))
  println("completed successfully")

File: 5_software_and_its_engineering/0_software_notations_and_tools/unified_discipline_of_computer_programming/programs_per_computational_model/3_stateful/arrays_and_strings/rotate-left.py
def new_index(index, rotations, arr_length):
    target_index = index - rotations
    abs_target_index = abs(target_index)
    if target_index < 0:
        return arr_length - abs_target_index
    else:
        return abs_target_index


def rotateLeft(d, arr):
    arr_length = len(arr)
    print(arr_length)
    new_array = [0 for _ in range(arr_length)]
    for i in range(arr_length):
        new_array[new_index(i, d, arr_length)] = arr[i]
    return new_array
